### PR TITLE
decouple message parsers from the core to enable later plugins

### PR DESCRIPTION
**Pull request by imzhenyu**
decouple message parsers from the core to enable later plugins such as raw or GRPC message parsers

### CODE REVIEW

File: src/core/tools/common/network.sim.cpp [hunk 1]
```diff
@@ -91,12 +91,8 @@ namespace dsn { namespace tools {
     {
         for (auto& msg : _sending_msgs)
         {
-            network_header_format hdr_format(NET_HDR_DSN);
-            bool r = header_type::header_type_to_format(msg->header->hdr_type, hdr_format);
-            dassert(r, "header_type_to_format(%s) failed", msg->header->hdr_type.debug_string().c_str());
-
             sim_network_provider* rnet = nullptr;
-            if (!s_switch[task_spec::get(msg->local_rpc_code)->rpc_call_channel][hdr_format].get(remote_address(), rnet))
+            if (!s_switch[task_spec::get(msg->local_rpc_code)->rpc_call_channel][NET_HDR_DSN].get(remote_address(), rnet))
```
Comment (qinzuoyan) [inline src/core/tools/common/network.sim.cpp]: the hdr_format should be determined by "msg->header->hdr_type", but not NET_HDR_DSN const value

File: src/core/tools/common/thrift_message_parser.h [hunk 2]
```diff
@@ -60,6 +60,10 @@ namespace dsn
     // response format:
     //     <total_len(int32)> <thrift_string> <thrift_message_begin> <body_data(bytes)> <thrift_message_end>
 
+# define THRIFT_HDR_SIG (*(uint32_t*)"THFT")
```
Comment (qinzuoyan) [inline src/core/tools/common/thrift_message_parser.h:63]: Why only define macro for THRIFT? or we can just use _(uint32_t_)"THFT" like others.

**Comment by imzhenyu** [comment]
the appveyor test failed because the computation resource is limited that the test cannot be finished within one hour.

File: src/core/tools/common/thrift_message_parser.h (hunk 2 above)
Comment (imzhenyu) [inline src/core/tools/common/thrift_message_parser.h:63]: There are multiple cases referencing this "THFT" in thrift-message-parser, and a macro helps make sure they are all consistent. For the other message parsers, they are referenced only once.

File: src/core/tools/common/network.sim.cpp (hunk 1 above)
Comment (imzhenyu) [inline src/core/tools/common/network.sim.cpp]: My bad~

File: src/core/core/rpc_message.cpp [hunk 3]
```diff
@@ -199,87 +199,11 @@ DSN_API void dsn_msg_get_options(
...
-        || hdr_type == hdr_type_http_options
-        || hdr_type == hdr_type_http_response
-        )
-        return DHT_HTTP;
-    else
-        return DHT_INVALID;
-}
-
 message_ex::message_ex()
     : header(nullptr), local_rpc_code(::dsn::TASK_CODE_INVALID), hdr_format(NET_HDR_DSN), send_retry_count(0),
```
Comment (qinzuoyan) [inline src/core/core/rpc_message.cpp:208]: we'd better init hdr_format value to "DSN_HDR_INVALID"

File: include/dsn/internal/message_parser.h [hunk 4]
```diff
@@ -130,6 +130,10 @@ namespace dsn
         // return buffer count used, which must be no more than the return value of get_buffer_count_on_send().
         // may be invoked for mutiple times if the message is reused for resending.
         virtual int get_buffers_on_send(message_ex* msg, /*out*/ send_buf* buffers) = 0;
+
+    public:
+        static network_header_format get_header_type(const char* bytes); // buffer size >= sizeof(uint32_t)
+        static std::string get_debug_string(const char* bytes);
     };
 
     class message_parser_manager : public utils::singleton<message_parser_manager>
```
Comment (qinzuoyan) [inline include/dsn/internal/message_parser.h:139]: we'd better init hdr_format value to "DSN_HDR_INVALID" in constructor of parser_factory_info.

**Pull request merged**